File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */
var lowestCommonAncestor = function(root, p, q) {
    if(!root) {
        return null;
    }
    if(root == p || root == q) {
        return root;
    }
    // Left and right traversal using recursion.
    const leftLCA = lowestCommonAncestor(root.left, p, q);
    const rightLCA = lowestCommonAncestor(root.right, p, q);
    
    if(leftLCA && rightLCA) {
        return root;
    }
    if(leftLCA) return leftLCA;
    
    return rightLCA;
};

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL) {
            return NULL;
        }
        
        if(root == p || root == q) {
            return root;
        }
        
        // Left and right traversal and calculation.
        TreeNode* LeftT = lowestCommonAncestor(root->left, p, q);
        TreeNode* RightT = lowestCommonAncestor(root->right, p, q);
        
        if(LeftT != NULL && RightT != NULL) {
            return root;
        }
        
        if(LeftT != NULL) {
            return LeftT;
        }
        
        return RightT;
    }
};